core/src/main/java/org/elasticsearch/search/slice/SliceBuilder.java

```diff
@@ -181,14 +189,54 @@ public int hashCode() {
         return Objects.hash(this.field, this.id, this.max);
     }
 
-    public Query toFilter(QueryShardContext context) {
+    public Query toFilter(QueryShardContext context, int shardId, int numShards) {
         final MappedFieldType type = context.fieldMapper(field);
         if (type == null) {
             throw new IllegalArgumentException("field " + field + " not found");
         }
+
         if (UidFieldMapper.NAME.equals(field)) {
-            return new TermsSliceQuery(field, id, max);
+            if (numShards == 1) {
+                return new TermsSliceQuery(field, id, max);
+            }
+
+            if (max >= numShards) {
+                // the number of slices is greater than the number of shards
+                // in such case we can reduce the number of requested shards by slice
+
+                // first we check if the slice is responsible of this shard
+                int targetShard = id % numShards;
+                if (targetShard != shardId) {
+                    // the shard is not part of this slice, we can skip it.
+                    return new MatchNoDocsQuery("this shard is not part of the slice");
+                }
+                // compute the number of slices where this shard appears
+                int numSlicesInShard = max / numShards;
+                int rest = max % numShards;
+                if (rest > targetShard) {
+                    numSlicesInShard ++;
+                }
+
+                if (numSlicesInShard == 1) {
+                    // this shard has only one slice so we must check all the documents
+                    return new MatchAllDocsQuery();
```

Comment:
Yep but it's more than max==numShards, we rewrite to match all whenever the shard appears only once in the remapped slice.